### DISCUSSION TITLE
Feature to store the window size and position on Windows and restore that position and size on next launch

### DISCUSSION THREAD

**Discussion by bmacombe** [CommunityToolkit/Maui#1777] [New Feature Discussions]
WinUI does not store the position and size of a window on close like UWP. So it is up to the application to implement its own mechanism.

I have implemented this feature in my Maui app and think it could be a useful feature too others.

The stored position is validated against the current display areas of the system in case the display configuration has changed since the Maui app was last started. If the stored position is no longer valid, the default WinUI position is used (basically do nothing to the window position on startup)

The basic logic flow for this feature is contained in a AppendToMapping on the WindowHandler.
1) A reference to the WinUI AppWindow object for the Maui app is obtained
2) Check for an existing stored position, the Preferences API from Maui is used to store the position. The user can supply their own key or a default key is used
3) If there is a stored position
 a) All the current DisplayAreas of the system are retrieved
 b) The stored position is checked against current DisplayAreas
 c) If the position is found to be within the current display areas (either fully contained or allow to cross displays at the users option) The AppWindow is moved and sized to the stored location. If no valid position is found, nothing is done
4) Wire an event to the AppWindow's Closing event, when fired the last location of the window is stored in the Preferences API.

All operations are wrapped in a try/catch to just ignore errors and the default behavior of just loading the default WinUI window location is used.

If approved I'm happy to provide an implementation.

**Comment by CliffAgius** [comment]
Not looked into it but would we want to do this for Desktop so cover Windows and Mac etc?

**Comment by TheCodeTraveler** [comment]
@pictos Mentioned in today's standup that the WinUIEx library already supports this feature: https://github.com/dotMorten/WinUIEx

**Reply by mrlacey** [reply]
Pointing people to Morten's existing library for this makes the most sense to me

**Reply by bmacombe** [reply]
Cool, I didn't know about that libaray, I agree no need to duplicate it.

**Reply by CliffAgius** [reply]
problem here @mrlacey is that Morten's existing library is not that discoverable...

**Reply by pictos** [reply]
For maui devs maybe... For windows devs I would say it's very discoverable, and I think it's normal since Morten's lib is platform specific and not a cross lib

**Reply by bmacombe** [reply]
The WinUIEx lib appears very useful and complete. I never tried it though as I already had my implementation complete and working well. I could be incorrect in my assumption, but I also doubt a lot of Maui devs are typical Windows Devs. I haven't touched a pure Windows app since Winforms (full framework Winforms) and WPF since I started using Xamarin Forms around 2015.

This line of code in my app sure seems easier for a Maui dev than https://dotmorten.github.io/WinUIEx/concepts/Maui.html (which I'm grateful has been added to the docs, I don't think I remember seeing that before!)
![image](https://github.com/user-attachments/assets/0606ac28-4c4a-4413-b453-b64b979eb034)

My whole thought around adding this feature to the toolkit was and still is, not to replicate anyone else's work or take away from it. But to make it easy for people coming from Xamarin Forms to get the behavior expected out of a XF UWP app (I know probably not a ton of those!) without a lot of effort.

**Reply by CliffAgius** [reply]
This is my thinking I have a WinUI build of a client project that is using this idea of positioning and sizing the windows when the app is opened but the code is by me as I wasn't aware of Morten's library until I see this thread.

However I do need to do the same for the app as the client wants a MacOS build soon and I am guessing they will want something similar as well.

I wonder how many other awesome libraries I am missing out on when writing code myself... Maybe we just need a list of awesome libraries...

**Reply by bmacombe** [reply]
Yes!, I wish I had known about Morten's library because I wouldn't have spent time trying to figure it out! I did a lot of searching on WinUI not storing windows position and never ran across WinUIEx in anything.

**Comment by CliffAgius** [comment]
Closed as answered, if closed in error please let us know to re-open.
